token-client-spring-demo/src/main/java/no/nav/security/token/support/demo/spring/client/BaseExampleClient.java

```diff
@@ -0,0 +1,41 @@
+package no.nav.security.token.support.demo.spring.client;
+
+import no.nav.security.token.support.oauth2.ClientConfigurationProperties;
+import no.nav.security.token.support.oauth2.client.OAuth2AccessTokenResponse;
+import no.nav.security.token.support.oauth2.client.OAuth2AccessTokenService;
+import org.springframework.boot.web.client.RestTemplateBuilder;
+import org.springframework.http.client.ClientHttpRequestInterceptor;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.Optional;
+
+public class BaseExampleClient {
+
+    protected final RestTemplate restTemplate;
+    private final ClientConfigurationProperties.ClientProperties clientProperties;
+    private final OAuth2AccessTokenService oAuth2AccessTokenService;
+
+    public BaseExampleClient(String clientConfigKey, RestTemplateBuilder restTemplateBuilder,
+                                   ClientConfigurationProperties clientConfigurationProperties,
+                                   OAuth2AccessTokenService oAuth2AccessTokenService) {
+
+        this.clientProperties = Optional.ofNullable(
+            clientConfigurationProperties.getClients().get(clientConfigKey))
+            .orElseThrow(() -> new RuntimeException("could not find oauth2 client config for key="+ clientConfigKey));
+        this.restTemplate = restTemplateBuilder
+            .rootUri(clientProperties.getResourceUrl().toString())
+            .interceptors(bearerTokenInterceptor())
+            .build();
+        this.oAuth2AccessTokenService = oAuth2AccessTokenService;
+
+    }
+
+    private ClientHttpRequestInterceptor bearerTokenInterceptor(){
+        return (request, body, execution) -> {
+            OAuth2AccessTokenResponse response =
+                oAuth2AccessTokenService.getAccessToken(clientProperties);
```

Comment:
Burde vi hatt et godt eksempel på hvordan gjøre fornuftig caching?
Tenker på det spesielt fordi akkurat her er det abstrahert bort hvorvidt det er "client_credentials" eller "on-behalf-of", mens caching-strategien jo vil være forskjellig for de to: i.e: For clientcred kan man bruke det ene tokenet man har i minnet (hvis varigheten er OK) uavhengig av hvem som spør, mens for on-behalf-of er det viktig å _kun_ benytte cache hvis man har en cache av OBO-token med nøkkel=innkommende-JWT. Og det er jo viktig at man sammenlikner med riktig token (hvis det potensielt skulle være flere), mens token-utplukkingen av det tokenet som benyttes i OBO-token-kallet mot utsteder ligger inne i OAuth2AccessTokenService.authenticatedJwtToken() i biblioteket.

Reply:
Godt poeng.. Mulig det ikke skulle være abstrahert bort? dvs. at authenticatedJwtToken er input til metoden, da kunne man ha brukt f.eks. @Cache annotering i og med at token da blir key i cache.. Da er det vanskelig å få til magic som mange har kommet til å forvente med Spring Boot og tildels token-support. Jeg er forsåvidt pro å være mer eksplisitt med litt mindre magi. Kanskje vi skulle diskutert denne litt muntlig også når du er på kontoret?

Reply:
Det kan vi godt. Men, ja.. kanskje det _enten_ burde abstraheres mindre, _eller_ - hvis man ønsker magi - kanskje faktisk mer? ;) At man legger også caching inn i OAuth2AccessTokenService, da eventuelt med config-mulighet for av/på og for expires-margin (og for max antall tokens i cache(?)) og med fornuftige defaults?

Reply:
@janolaveide Har du tid til å ta en rask titt? Og evt. gjøre deg opp en mening angående det vi diskuterer over her?

Reply:
dettte så litt mer detaljert ut enn det jeg kan si noe fornuftig om i farten, gitt .  :-)

Reply:
@royag-nav @tronghn Nå med caching. har todo på opt-in og tester.